russh/src/channels/mod.rs

```diff
@@ -111,7 +118,7 @@ pub struct Channel<Send: From<(ChannelId, ChannelMsg)>> {
     pub(crate) sender: Sender<Send>,
     pub(crate) receiver: UnboundedReceiver<ChannelMsg>,
     pub(crate) max_packet_size: u32,
-    pub(crate) window_size: u32,
+    pub(crate) window_size: Arc<Mutex<u32>>,
```

Comment:
```suggestion
    pub(crate) window_size: Arc<AtomicU32>,
```

Possible to do this instead to avoid `Mutex`?

Reply:
Hi there, thanks for the suggestion, I though of using Atomics for the `window_size`, however there are two reasons this is not suitable for this case:
- Atomics operations might be blocking which is strongly discouraged in an _async_ context, because it might block the event loop and there are no non-blocking counterparts of those operations (ala `Mutex::try_lock`);
- We _must_ hold the Mutex in the following code to ensure there are no race-conditions because we first fetch and then update the value, this means no other thread can act on the `window_size` and lead to an overshoot with two threads sending data causing the window to be exceeded.
```rust
let mut window_size = match self.window_size.try_lock() {
    Ok(window_size) => window_size, 
    Err(_) => {
        cx.waker().wake_by_ref();
        return Poll::Pending;
    }
};

let writable = self.max_packet_size.min(*window_size).min(buf.len() as u32) as usize;
if writable == 0 {
    cx.waker().wake_by_ref();
    return Poll::Pending;
}

let mut data = CryptoVec::new_zeroed(writable);
#[allow(clippy::indexing_slicing)] // Clamped to maximum `buf.len()` with `.min`
data.copy_from_slice(&buf[..writable]);
data.resize(writable);

*window_size -= writable as u32;
drop(window_size);
```